File: src/ProfilerExtension.java
package org.nlogo.extensions.profiler;

import org.nlogo.core.Syntax;
import org.nlogo.core.SyntaxJ;
import org.nlogo.api.Context;
import org.nlogo.api.Command;
import org.nlogo.api.Reporter;
import org.nlogo.api.Argument;
import org.nlogo.api.Context;
import org.nlogo.api.ExtensionException;
import org.nlogo.api.LogoException;
import org.nlogo.nvm.ExtensionContext;
import org.nlogo.nvm.Tracer;
import org.nlogo.nvm.ExtensionContext;

public class ProfilerExtension extends org.nlogo.api.DefaultClassManager {
  public void load(org.nlogo.api.PrimitiveManager primManager) {

    primManager.addPrimitive("start", new ProfilerStart());
    primManager.addPrimitive("stop", new ProfilerStop());
    primManager.addPrimitive("reset", new ProfilerReset());
    primManager.addPrimitive("report", new ProfilerReport());
    primManager.addPrimitive("calls", new ProfilerProcedureCalls());
    primManager.addPrimitive("exclusive-time", new ProfilerProcedureExclusiveTime());
    primManager.addPrimitive("inclusive-time", new ProfilerProcedureInclusiveTime());
  }

  public void runOnce(org.nlogo.api.ExtensionManager em)
      throws org.nlogo.api.ExtensionException {
    Tracer tracer = new QuickTracer();
    ((org.nlogo.workspace.ExtensionManager) em).workspace().setProfilingTracer(tracer);
    // we disable it once it is installed, so that
    // we don't start collecting profiling data until
    // ProfilerStart is called.  -- CLB
    tracer.disable();
  }

  public void unload(org.nlogo.api.ExtensionManager em) {
    ((org.nlogo.workspace.ExtensionManager) em).workspace().setProfilingTracer(null);
  }

  public static class ProfilerStart implements Command {
    public Syntax getSyntax() {
      return SyntaxJ.commandSyntax();
    }

    public void perform(Argument args[], Context context) throws ExtensionException {

      if (!org.nlogo.api.Version$.MODULE$.useGenerator()) {
        throw new ExtensionException
            ("The profiler extension requires the NetLogo bytecode generator, "
                + "which is currently turned off. See the org.nlogo.noGenerator property.");
      }
      ((ExtensionContext) context).workspace().profilingTracer().enable();
    }
  }

  public static class ProfilerStop implements Command {
    public Syntax getSyntax() {
      return SyntaxJ.commandSyntax();
    }

    public void perform(Argument args[], Context context) throws ExtensionException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        tracer.disable();
      }
    }
  }

  public static class ProfilerReset implements Command {
    public Syntax getSyntax() {
      return SyntaxJ.commandSyntax();
    }

    public void perform(Argument args[], Context context) throws ExtensionException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        tracer.reset();
      }
    }

  }

  public static class ProfilerReport implements Reporter {
    public Syntax getSyntax() {
      return SyntaxJ.reporterSyntax(Syntax.ListType());
    }

    public Object report(Argument args[], Context context) throws ExtensionException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        java.io.ByteArrayOutputStream outArray = new java.io.ByteArrayOutputStream();
        java.io.PrintStream out = new java.io.PrintStream(outArray);
        tracer.dump(out);
        return outArray.toString();
      }
      return "";
    }
  }

  public static class ProfilerProcedureCalls implements Reporter {
    public Syntax getSyntax() {
      return SyntaxJ.reporterSyntax
          (new int[]{Syntax.StringType()},
              Syntax.NumberType());
    }

    public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        String arg0 = args[0].getString().toUpperCase();
        return Double.valueOf(tracer.calls(arg0));
      }
      return Double.valueOf(0);
    }
  }

  public static class ProfilerProcedureExclusiveTime implements Reporter {
    public Syntax getSyntax() {
      return SyntaxJ.reporterSyntax
          (new int[]{Syntax.StringType()},
              Syntax.NumberType());
    }

    public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        String arg0 = args[0].getString().toUpperCase();
        return Double.valueOf(tracer.exclusiveTime(arg0) / 1000000.0);
      }
      return Double.valueOf(0);
    }
  }

  public static class ProfilerProcedureInclusiveTime implements Reporter {
    public Syntax getSyntax() {
      return SyntaxJ.reporterSyntax
          (new int[]{Syntax.StringType()},
              Syntax.NumberType());
    }

    public Object report(Argument args[], Context context) throws ExtensionException, LogoException {
      Tracer tracer = ((ExtensionContext) context).workspace().profilingTracer();
      if (tracer != null) {
        String arg0 = args[0].getString().toUpperCase();
        return Double.valueOf(tracer.inclusiveTime(arg0) / 1000000.0);
      }
      return Double.valueOf(0);
    }
  }
}

File: src/scala/ProfilerExtension.scala
import java.io.{ ByteArrayOutputStream, PrintStream }
import java.util.Arrays
import java.util.stream.Stream

import org.nlogo.api.{ Argument, Context, Command, DefaultClassManager, ExtensionException, ExtensionManager
                     , LogoException, PrimitiveManager, Reporter, Version }
import org.nlogo.core.{ LogoList, Syntax, SyntaxJ }
import Syntax.{ commandSyntax, ListType, NumberType, reporterSyntax, StringType }
import org.nlogo.nvm.{ ExtensionContext, Tracer, Workspace }
import org.nlogo.workspace.{ ExtensionManager => WSEM }

class ProfilerExtension extends DefaultClassManager {

  override def load(primManager: PrimitiveManager): Unit = {
    primManager.addPrimitive("start",          ProfilerStart)
    primManager.addPrimitive("stop",           ProfilerStop)
    primManager.addPrimitive("reset",          ProfilerReset)
    primManager.addPrimitive("report",         ProfilerReport)
    primManager.addPrimitive("data",           ProfilerData)
    primManager.addPrimitive("calls",          ProfilerProcedureCalls)
    primManager.addPrimitive("exclusive-time", ProfilerProcedureExclusiveTime)
    primManager.addPrimitive("inclusive-time", ProfilerProcedureInclusiveTime)
  }

  override def runOnce(em: ExtensionManager): Unit = {
    val tracer = new QuickTracer()
    em.asInstanceOf[WSEM].workspace.setProfilingTracer(tracer)
    // we disable it once it is installed, so that
    // we don't start collecting profiling data until
    // ProfilerStart is called.  -- CLB
    tracer.disable()
  }

  override def unload(em: ExtensionManager): Unit = {
    em.asInstanceOf[WSEM].workspace.setProfilingTracer(null)
  }

  private object ProfilerStart extends Command {
    override def getSyntax = commandSyntax()
    override def perform(args: Array[Argument], context: Context): Unit = {
      if (!Version.useGenerator) {
        val msg = """The profiler extension requires the NetLogo bytecode generator, which is currently turned off.
                    |See the org.nlogo.noGenerator property for more info.""".stripMargin
        throw new ExtensionException(msg)
      }
      tracerOpt(context).foreach {
        _.enable()
      }
    }
  }

  private object ProfilerStop extends Command {
    override def getSyntax = commandSyntax()
    override def perform(args: Array[Argument], context: Context): Unit = {
      tracerOpt(context).foreach {
        _.disable()
      }
    }
  }

  private object ProfilerReset extends Command {
    override def getSyntax = SyntaxJ.commandSyntax()
    override def perform(args: Array[Argument], context: Context): Unit = {
      tracerOpt(context).foreach {
        _.reset()
      }
    }

  }

  private object ProfilerReport extends Reporter {
    override def getSyntax = reporterSyntax(ret = StringType)
    override def report(args: Array[Argument], context: Context) = {
      tracerOpt(context).fold("") {
        tracer =>
          val outArray = new ByteArrayOutputStream()
          tracer.dump(new PrintStream(outArray))
          outArray.toString()
      }
    }
  }

  private object ProfilerData extends Reporter {
    override def getSyntax = reporterSyntax(ret = ListType)
    override def report(args: Array[Argument], context: Context) = {
      import scala.jdk.CollectionConverters.SetHasAsScala
      tracerOpt(context).fold(LogoList.Empty) {
        tracer =>
          val headers = LogoList("procedure", "calls", "inclusive_time", "exclusive_time")
          val rows    =
            tracer.procedureNames().asScala.toSeq.map {
              procedureName =>
                LogoList( procedureName
                        , Double.box(tracer.calls(procedureName).toDouble)
                        , Double.box(tracer.inclusiveTime(procedureName) / 1e6)
                        , Double.box(tracer.exclusiveTime(procedureName) / 1e6)
                        )
            }
          LogoList((headers +: rows)*)
      }
      //final Workspace workspace = ((ExtensionContext) context).workspace()
      //final QuickTracer tracer = ((QuickTracer) workspace.profilingTracer())
    }
  }

  private object ProfilerProcedureCalls extends Reporter {
    override def getSyntax = reporterSyntax(right = List(StringType), ret = NumberType)
    override def report(args: Array[Argument], context: Context) = {
      val name = args(0).getString.toUpperCase
      tracerOpt(context).fold(Double.box(0)) {
        _.calls(name).toDouble
      }
    }
  }

  private object ProfilerProcedureExclusiveTime extends Reporter {
    override def getSyntax = reporterSyntax(right = List(StringType), ret = NumberType)
    override def report(args: Array[Argument], context: Context) = {
      val name = args(0).getString.toUpperCase
      tracerOpt(context).fold(Double.box(0)) {
        _.exclusiveTime(name) / 1e6
      }
    }
  }

  private object ProfilerProcedureInclusiveTime extends Reporter {
    override def getSyntax = reporterSyntax(right = List(StringType), ret = NumberType)
    override def report(args: Array[Argument], context: Context) = {
      val name = args(0).getString.toUpperCase
      tracerOpt(context).fold(Double.box(0)) {
        _.inclusiveTime(name) / 1e6
      }
    }
  }

  private def tracerOpt(context: Context): Option[Tracer] =
    Option(context.asInstanceOf[ExtensionContext].workspace.profilingTracer)

}
